File: src/main/java/br/unifor/costify/application/usecase/CalculateRecipeCostUseCase.java
package br.unifor.costify.application.usecase;

import br.unifor.costify.application.contracts.IngredientRepository;
import br.unifor.costify.application.contracts.RecipeRepository;
import br.unifor.costify.application.dto.response.RecipeCostDto;
import br.unifor.costify.application.errors.IngredientNotFoundException;
import br.unifor.costify.application.errors.RecipeNotFoundException;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.entity.Recipe;
import br.unifor.costify.domain.service.RecipeCostCalculationService;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.RecipeCost;
import br.unifor.costify.domain.valueobject.RecipeIngredient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class CalculateRecipeCostUseCase {
    
    private final RecipeRepository recipeRepository;
    private final IngredientRepository ingredientRepository;
    private final RecipeCostCalculationService costCalculationService;
    
    public CalculateRecipeCostUseCase(
            RecipeRepository recipeRepository,
            IngredientRepository ingredientRepository,
            RecipeCostCalculationService costCalculationService) {
        this.recipeRepository = recipeRepository;
        this.ingredientRepository = ingredientRepository;
        this.costCalculationService = costCalculationService;
    }
    
    public RecipeCostDto execute(String recipeId) {
        Id id = Id.of(recipeId);
        
        Optional<Recipe> recipeOpt = recipeRepository.findById(id);
        if (recipeOpt.isEmpty()) {
            throw new RecipeNotFoundException("Recipe not found with ID: " + recipeId);
        }
        
        Recipe recipe = recipeOpt.get();
        Map<Id, Ingredient> ingredientMap = loadIngredients(recipe.getIngredients());
        
        RecipeCost recipeCost = costCalculationService.calculateCost(recipe, ingredientMap);
        
        return RecipeCostDto.fromDomain(recipeCost);
    }
    
    private Map<Id, Ingredient> loadIngredients(List<RecipeIngredient> recipeIngredients) {
        Map<Id, Ingredient> ingredientMap = new HashMap<>();
        
        for (RecipeIngredient recipeIngredient : recipeIngredients) {
            Id ingredientId = recipeIngredient.getIngredientId();
            
            Optional<Ingredient> ingredientOpt = ingredientRepository.findById(ingredientId);
            if (ingredientOpt.isEmpty()) {
                throw new IngredientNotFoundException("Ingredient not found with ID: " + ingredientId);
            }
            
            ingredientMap.put(ingredientId, ingredientOpt.get());
        }
        
        return ingredientMap;
    }
}

File: src/application/usecase/CalculateRecipeCostUseCase.ts
import type { DecimalProvider } from '@domain/contracts/DecimalProvider.js'
import { RecipeCostCalculationService } from '@domain/services/RecipeCostCalculationService.js'
import { Id } from '@domain/valueobjects/Id.js'
import type { IngredientRepository } from '../contracts/IngredientRepository.js'
import type { RecipeRepository } from '../contracts/RecipeRepository.js'
import type { CalculateRecipeCostCommand } from '../dto/command/CalculateRecipeCostCommand.js'
import type { RecipeCostDto } from '../dto/response/RecipeCostDto.js'
import { IngredientNotFoundException } from '../errors/IngredientNotFoundException.js'
import { RecipeNotFoundException } from '../errors/RecipeNotFoundException.js'
import { RecipeMapper } from '../mapper/RecipeMapper.js'

export class CalculateRecipeCostUseCase {
  private readonly costCalculationService: RecipeCostCalculationService

  constructor(
    private readonly recipeRepository: RecipeRepository,
    private readonly ingredientRepository: IngredientRepository,
    private readonly decimalProvider: DecimalProvider
  ) {
    this.costCalculationService = new RecipeCostCalculationService(this.decimalProvider)
  }

  async execute(command: CalculateRecipeCostCommand): Promise<RecipeCostDto> {
    const recipeId = new Id(command.recipeId)

    // Find recipe
    const recipe = await this.recipeRepository.findById(recipeId)
    if (!recipe) {
      throw new RecipeNotFoundException(command.recipeId)
    }

    // Get all required ingredients
    const ingredients = []
    for (const recipeIngredient of recipe.getIngredients()) {
      const ingredient = await this.ingredientRepository.findById(
        recipeIngredient.getIngredientId()
      )
      if (!ingredient) {
        throw new IngredientNotFoundException(recipeIngredient.getIngredientId().getValue())
      }
      ingredients.push(ingredient)
    }

    // Calculate recipe cost
    const recipeCost = this.costCalculationService.calculateRecipeCost(recipe, ingredients)

    // Update recipe with calculated cost (optional - for caching purposes)
    recipe.updateTotalCost(recipeCost.getTotalCost())
    await this.recipeRepository.update(recipe)

    // Return cost DTO
    return RecipeMapper.recipeCostToDto(recipeCost)
  }
}
